Update to clang format 19 support

hjmjohnson opened the pull request:
Resolves: https://github.com/InsightSoftwareConsortium/ITK/issues/4565

Update to newer version of clang-format.

The previous clang-format version 8 does not compile on arm64 mac.  There is limited support for clang-format binaries before version 10.

This is a gateway toward greatly simplifying the support for pre-commit checking using the python pre-commit infrastructure.  See: https://github.com/InsightSoftwareConsortium/ITK/pull/5013


## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

hjmjohnson commented:
FYI:  Github actions is having issues:  https://github.com/actions/runner/issues/3611

Local macos builds occur properly (my main development environment is macos).

blowekamp commented:
The end goal is to have precommit manage the clang format version, and remove our code with build and manages the clang format version?

hjmjohnson commented:
> The end goal is to have precommit manage the clang format version, and remove our code with build and manages the clang format version?

@blowekamp   Yes! The pre-commit version is much easier to manage and covers a wider range of platforms. Using pre-commit will remove much of the complicated cmake code.

I made this PR as a stepping stone to decouple the pre-commit considerations from updating the clang-format version.

dzenanz commented on Modules/Core/Transform/include/itkBSplineTransform.hxx:
```diff
@@ -440,8 +440,7 @@ BSplineTransform<TParametersValueType, VDimension, VSplineOrder>::SetCoefficient
   {
-    itkExceptionMacro("SetCoefficientImage() requires that an array of "
-                      << "correctly sized images be supplied.");
+    itkExceptionMacro("SetCoefficientImage() requires that an array of " << "correctly sized images be supplied.");
```
`" << "` in the middle of the string can be removed now. Is there an automated way to accomplish that?

hjmjohnson replied:
Done.

N-Dekker replied:
Removing " << " is a real code change. It's not just a format change. Although it's a change that I like, I would prefer such a change in a separate PR.

dzenanz replied:
Follow-up in PR #5052.

dzenanz commented on Modules/Filtering/BinaryMathematicalMorphology/include/itkBinaryErodeImageFilter.hxx:
```diff
@@ -324,9 +324,9 @@ BinaryErodeImageFilter<TInputImage, TOutputImage, TKernel>::GenerateData()
-      }       // if( bIsOnContour )
+          } // for (i = 0; i < neighborhoodSize; ++i)
+        } // while ( !propagQueue.empty() )
+      } // if( bIsOnContour )
```
New alignment is better

hjmjohnson replied:
yes.

dzenanz commented on Modules/Filtering/DisplacementField/include/itkInvertDisplacementFieldImageFilter.hxx:
```diff
@@ -109,7 +109,7 @@ InvertDisplacementFieldImageFilter<TInputImage, TOutputImage>::GenerateData()
   float oldProgress = 0.0f;
 
-  while (iteration++ < this->m_MaximumNumberOfIterations && this->m_MaxErrorNorm > this->m_MaxErrorToleranceThreshold &&
+  while (iteration++<this->m_MaximumNumberOfIterations && this->m_MaxErrorNorm> this->m_MaxErrorToleranceThreshold &&
```
Strange re-spacing of comparison operators

hjmjohnson replied:
Agreed, I'll look into why that occurred.

hjmjohnson replied:
Adding parenthesis around conditionals makes this much easier to read, and fixes the formatting issue.

dzenanz commented on .clang-format:
```diff
@@ -219,7 +219,14 @@ PenaltyIndentedWhitespace: 0
+QualifierOrder:
+  - static
+  - inline
+  - friend
```
`friend` is the rarest qualifier by far. I would prefer it be the first.

hjmjohnson replied:
Easy fix will do for next time.

dzenanz reviewed (commented):
Thanks for doing this!

hjmjohnson commented on .clang-format:
```diff
@@ -10,139 +10,273 @@
+  AlignFunctionPointers: false
+  PadOperators:    false
+AlignConsecutiveDeclarations:
+  Enabled:         true
```
Can we delay this change to a later commit?  There have been strong feelings both ways on this topic. 

I happen to agree with you, but do not want to conflate two issues.

dzenanz replied:
Yes!

Making this change (if we agree to it) should be done in a separate PR.

seanm replied:
FWIW I also dislike `AlignConsecutiveDeclarations`. It pollutes blame, it makes changes to lines that aren't really changing, complicating code review, etc. etc.

hjmjohnson commented on Examples/.clang-format:
```diff
@@ -10,142 +10,300 @@
+  AlignFunctionPointers: false
+  PadOperators:    false
+AlignConsecutiveDeclarations:
+  Enabled:         true
```
Can we delay this change to a later commit?  There have been strong feelings both ways on this topic. 

I happen to agree with you, but do not want to conflate two issues.

hjmjohnson commented on Modules/Core/Common/include/itkMetaProgrammingLibrary.h:
```diff
@@ -56,7 +56,7 @@ struct FalseType
   using Type = FalseType;
   static constexpr ValueType Value = false;
-                             operator ValueType() { return Value; }
+  operator ValueType() { return Value; }
```
?? don't know?  previous clang-format 8 was also confused by this line.

I will manually expand and see if it sticks.

hjmjohnson replied:
setting .clang-format  "AllowShortFunctionsOnASingleLine: All" minimizes differentces with clang-format version 8.    130 fewer files changed any content.

========
I prefer 
AllowShortFunctionsOnASingleLine: None which expands all short functions, but that adds many more differences over the previous clang version 8.

hjmjohnson commented on Modules/Core/Common/test/itkAnatomicalOrientationGTest.cxx:
```diff
@@ -169,8 +169,8 @@ TEST(AnatomicalOrientation, ConvertDirectionToPositiveEnum)
   const itk::SpacePrecisionType                data[] = { 0.5986634407395047, 0.22716302314740483, -0.768113953548866,
-                                           0.5627936241740271, 0.563067040943212,   0.6051601804419384,
-                                           0.5699696670095713, -0.794576911518317,  0.20924175102261847 };
+                                                          0.5627936241740271, 0.563067040943212,   0.6051601804419384,
```
Me too.

hjmjohnson commented:
@dzenanz @blowekamp I have addressed all the issues that you pointed out earlier.  I found a setting that minimized the differences with clang version 8 with ~130 fewer changed files.

I'd love to get this PR to make progress toward acceptance because it is a key roadblock for evaluating the next set of build improvements (aka pre-commit).

blowekamp reviewed (approved):
Thank you for moving to a much needed newer version of clang-formatting. It is good to separate the stye preferences from version changes as best we can. Additionally, separating the code formatting changes from the introduction of pre-commit usage is good. Thank you for moving this forwards.

N-Dekker commented on Modules/Bridge/VTK/src/itkVTKImageExportBase.cxx:
```diff
@@ -23,10 +23,7 @@ namespace itk
-{
-  m_LastPipelineMTime = 0;
-}
+VTKImageExportBase::VTKImageExportBase() { m_LastPipelineMTime = 0; }
```
Honestly I like the original better, having the `m_LastPipelineMTime = 0;` on its own line. (No show stopper to me, just a matter of taste, I guess.)

blowekamp replied:
Please don't regress this PR in to a line by line critic of style. The goal is to update to a newer version clang format matching the current style as close as possible.

N-Dekker replied:
@blowekamp Don't worry, I don't do 1000+ lines of code line by line. 😸 But it's good to know what kind of formatting changes we might expect from upgrading clang-format.

hjmjohnson replied:
I investigated options for minimizing code changes.  

I settled on changes that minimized differences with what clang-8 was doing:
AllowShortFunctionsOnASingleLine: All

I definitely like
 AllowShortFunctionsOnASingleLine: None 
much better, and that would make this line like it was. but other files would have changed.

I have a comment in the .clang-format file and plan to introduce the new option soon (this weekend :).

Hans

N-Dekker commented on Modules/Core/Common/include/itkAutoPointer.h:
```diff
@@ -73,7 +73,11 @@ class AutoPointer
 
   /** Overload operator ->.  */
-  ObjectType * operator->() const { return m_Pointer; }
+  ObjectType *
+  operator->() const
+  {
+    return m_Pointer;
+  }
```
Interestingly, here it goes the other way (compared to the case of the `m_LastPipelineMTime = 0;` statement). I like having the `return m_Pointer;` statement on its own line.  👍

N-Dekker commented on Modules/Core/Common/include/itkImageScanlineConstIterator.h:
```diff
@@ -265,7 +265,7 @@ class ITK_TEMPLATE_EXPORT ImageScanlineConstIterator : public ImageConstIterator
 template <typename TImage>
 ImageScanlineConstIterator(SmartPointer<TImage>, const typename TImage::RegionType &)
-  ->ImageScanlineConstIterator<std::remove_const_t<TImage>>;
+  -> ImageScanlineConstIterator<std::remove_const_t<TImage>>;
```
Glad to see that the new clang format "understands" the syntax of those CTAD deduction guides 👍

N-Dekker commented:
@hjmjohnson Thanks, really great! I'm very glad to see such a major clang-format upgrade! 

I'm now going to try your PR locally. Don't know if it's necessary to wait until it works on my Windows laptop...

N-Dekker commented:
@hjmjohnson  Dummy question: I just pulled the master branch and reran CMake. An attempt to commit locally tells me:

```
clang-format version 19.1.4 is required
Set the path the clang-format 19.1.4 executable with
  git config clangFormat.binary /path/to/clang-format
```
What should I do? I don't see the clang-format 19.1.4 executable on my disk. My "Bin\ITK\ClangFormat-prefix\src\ClangFormat" still has an old clang-format.exe

dzenanz commented:
@N-Dekker From this commit 0b1fabd319697a61e10602804f4eb02e455ac6a1 it can be deduced to get Windows binary from [this link](https://data.kitware.com/api/v1/file/hashsum/sha512/0e6d7176927e52acf081f47ff7cda1cd8d093931c80889f3765a3dca673633f8c1326b9977c5bf79f6d32e98632cb83d91e871aca18081477f70a715059454cf/download).
